✨ 增加 html to image 的缓存

chuangbo opened the pull request:
问题

用户短时间内发送多条重复命令，返回的图片未发生改变。在这样的情况下，希望派蒙可以提高回复速度。

方案

* 增加 html to image 缓存
* 对 template_service.render 进行封装，管理缓存逻辑
* cache key 为 html 的 sha256
* cache value 为 reply_photo 后 telegram 返回的 file_id
* 存入 redis，并设置合理的 ttl

## 该 PR 相关 Issue / Involved issue

Close #189

## 描述 / Description

```
见 #189 讨论
```

## 更改检查表 / Change Checklist
  
- [ ] 有插件命令更新
  - [ ] 已更新 bot 帮助文件
- [ ] 有流程交互
  - [ ] 指引明确
  - [ ] 可以退出
- [ ] 会触发频率限制
  - [ ] 有对应的限制措施
- [ ] 需要用户输入数据
  - [ ] 验证用户数据
  - [ ] 如果是文件，检查了文件大小
  - [ ] 对保存的数据再次进行了验证
- [ ] 添加了新的依赖包
- [x] 测试
  - [x] 本地通过了测试
  - [x] CI 通过了测试

## 说明 / Note

chuangbo committed:
feat: 增加 html to image 的缓存

问题

用户短时间内发送多条重复命令，返回的图片未发生改变。在这样的情况下，希望派蒙可以提高回复速度。

方案

* 增加 html to image 缓存
* 对 template_service.render 进行封装，管理缓存逻辑
* cache key 为 html 的 sha256
* cache value 为 reply_photo 后 telegram 返回的 file_id
* 存入 redis，并设置合理的 ttl

luoshuijs commented on core/template/cache.py:
```diff
@@ -26,3 +27,26 @@ async def set_data(self, key: str, data: Any, ttl: int = 8 * 60 * 60):
+        if data:
+            return data.decode()
+
+    async def set_data(self, html: str, file_id: str, ttl: int = 60 * 60):
```
我觉得默认 `TTL` 1小时太短了吧，至少是大于一周甚至一个月，因为有些数据根本真的1个月都不会发生变化，比如说角色列表

luoshuijs replied:
~~有的人已经非到几个版本角色列表动都不动了，还有屯屯鼠的那种~~

luoshuijs commented on core/template/services.py:
```diff
@@ -101,7 +116,35 @@ async def render(
+        return RenderResult(html=html, photo=png_data, cache=self.html_to_file_id_cache)
+
+
+class RenderResult:
```
还有个 `reply_document` 需要添加

luoshuijs commented on core/template/services.py:
```diff
@@ -101,7 +116,35 @@ async def render(
+        self.photo = photo
+        self._cache = cache
+
+    async def reply_photo(self, message: Message, *args, **kwargs):
```
给 `reply_photo` 添加 `ttl` 可选参数，包括后面的

luoshuijs committed:
✨ 支持 `media_group`

luoshuijs committed:
🎨 规范目录结构

luoshuijs committed:
Merge branch 'main' into feature/render-cache

chuangbo committed:
🐛 cache key 增加 file_type

reply photo 方式返回的 file_id 不能用于 reply_document

chuangbo committed:
🐛 修复循环引用

luoshuijs committed:
🎨 添加类型判断和自定义TTL等

luoshuijs committed:
Merge branch 'main' into feature/render-cache

luoshuijs committed:
🎨 修改函数TTL和缓存的QNAME

luoshuijs committed:
🐛 修复BUG

luoshuijs committed:
🎨 修改部分插件并添加 `edit_media`

luoshuijs committed:
🎨 修改 `HELP` 插件的TTL时间

chuangbo committed:
🎨 abyss 插件使用 html to file_id cache 以及并发渲染

chuangbo committed:
🐛 修复武器查询

ttl 是 render 的参数

chuangbo committed:
🎨 去掉 render_group 和 InputRenderData 并修改 abyss 的实现

render_group 和 InputRenderData 用起来比较繁琐，初始化 pydantic 需要每一个
key 都写上，绕了一圈。感觉在调用方自己并发 gather 然后再拼一个 RenderGroupResult
再 reply_media_group 就足够了。

karakoo committed:
⚡️ 提升 `avatars-list` 的性能
并发获取角色详情

Signed-off-by: Karako <karakohear@gmail.com>

chuangbo committed:
Merge remote-tracking branch 'origin/main' into feature/render-cache

chuangbo committed:
🎨 daily_material 插件使用 render cache

* 去掉渲染结果中的具体时间日期，提高缓存命中率

luoshuijs committed:
Merge branch 'main' into feature/render-cache

luoshuijs committed:
Merge branch 'main' into feature/render-cache

luoshuijs committed:
🐛 修复文章定时抓取时返回全部文章而不是新文档的问题

chuangbo committed:
🎨 cache_file_id 直接使用 self.ttl

chuangbo committed:
🐛 文档 file_id 在 reply.document.file_id

chuangbo committed:
🎨 /avatars full 使用 render cache

* 顺便去掉不必要的 `asyncio.create_task`

chuangbo committed:
🎨 gacha_log gacha_count 使用 render cache

luoshuijs committed:
🐛 修改部分插件TTL缓存时间

luoshuijs committed:
Merge branch 'main' into feature/render-cache

omg-xtao committed:
仅生产环境返回缓存
